src/staking-router/staking-router.service.ts

```diff
@@ -18,39 +19,88 @@ export class StakingRouterService {
     protected readonly keysApiService: KeysApiService,
   ) {}
 
-  /**
-   * Return staking module data and block information
-   */
-  public async getStakingModulesData(): Promise<{
-    stakingModulesData: StakingModuleData[];
-    blockHash: string;
-    blockNumber: number;
-  }> {
+  protected stakingRouterCache: Record<number, StakingModuleData> = {};
+
+  async getOperatorsAndModules() {
     const { data: operatorsByModules, meta: operatorsMeta } =
       await this.keysApiService.getOperatorListWithModule();
 
+    return { data: operatorsByModules, meta: operatorsMeta };
+  }
+
+  async fetchUnusedKeys(moduleAddresses: string[]) {
     const { data: unusedKeys, meta: unusedKeysMeta } =
-      await this.keysApiService.getUnusedKeys();
+      await this.keysApiService.getUnusedKeys(moduleAddresses);
 
-    const blockHash = operatorsMeta.elBlockSnapshot.blockHash;
-    const blockNumber = operatorsMeta.elBlockSnapshot.blockNumber;
+    return { data: unusedKeys, meta: unusedKeysMeta };
+  }
+
+  /**
+   * Return staking module data and block information
+   */
+  async getStakingModulesData(
+    data: GroupedByModuleOperatorListResponse,
+  ): Promise<StakingModuleData[]> {
+    const { data: operatorsByModules, meta: operatorsMeta } = data;
+
+    const cachedData: StakingModuleData[] = [];
+    const outdatedModuleAddresses: string[] = [];
+
+    // if module was deleted we will not include data in final result
+    // TODO: add test on delete
+    operatorsByModules.forEach(({ module: stakingModuleData }) => {
+      if (this.isDataOutdated(stakingModuleData)) {
+        outdatedModuleAddresses.push(stakingModuleData.stakingModuleAddress);
+        return;
+      }
+      cachedData.push(this.stakingRouterCache[stakingModuleData.id]);
+      return;
+    });
+
+    if (!outdatedModuleAddresses.length) return cachedData;
+
+    const { data: unusedKeys, meta: unusedKeysMeta } =
+      await this.fetchUnusedKeys(outdatedModuleAddresses);
 
     this.isEqualLastChangedBlockHash(
       operatorsMeta.elBlockSnapshot.lastChangedBlockHash,
       unusedKeysMeta.elBlockSnapshot.lastChangedBlockHash,
     );
 
-    const stakingModulesData = operatorsByModules.map(
-      ({ operators, module: stakingModule }) =>
+    const actualizedData = operatorsByModules
```

Comment:
I would suggest greatly simplifying this code, leaving only the keys to be cached